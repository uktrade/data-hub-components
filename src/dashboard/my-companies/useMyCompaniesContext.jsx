import { orderBy, pick } from 'lodash'
import { useReducer } from 'react'
import createUseContext from 'constate'
import { FILTER_CHANGE, LIST_CHANGE, ORDER_CHANGE } from './constants'

export const filterCompanyName = (companies, filterText) =>
  filterText.length
    ? companies.filter((c) =>
        c.company.name.toLowerCase().includes(filterText.toLowerCase())
      )
    : companies

const defaultState = {
  lists: [],
  selectedIdx: 0,
  sortBy: 'recent',
  filter: '',
}

export const reducer = (state, { type, ...action }) => {
  switch (type) {
    case LIST_CHANGE:
      return { ...state, selectedIdx: action.idx }
    case FILTER_CHANGE:
      return { ...state, filter: action.filter }
    case ORDER_CHANGE:
      return { ...state, sortBy: action.sortBy }
    default:
      return state
  }
}

// We are unpacking children here just to remove them from initialState.
const useMyCompaniesContext = createUseContext(
  ({ children, editListsLinkProps, ...rest }) => {
    const initialState = pick(rest, [
      'lists',
      'selectedIdx',
      'sortBy',
      'filter',
    ])
    const [state, dispatch] = useReducer(reducer, {
      ...defaultState,
      ...initialState,
      // Reducer assumes the list of company lists is sorted.
      lists: orderBy(initialState.lists, 'name'),
    })

    return {
      state,
      dispatch,
      editListsLinkProps,
    }
  }
)

export default useMyCompaniesContext
