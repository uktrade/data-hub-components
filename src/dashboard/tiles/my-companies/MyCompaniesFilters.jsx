import React, { useState } from 'react'

function MyCompaniesFilters({ initialCompanies, companies, onFilterUpdate }) {
  const [filterCompanyName, setFilterCompanyName] = useState('')

  function filterByName(e) {
    const companyName = e.target.value
    setFilterCompanyName(companyName)

    const filteredCompanies = initialCompanies.filter(
      c => c.companyName.toUpperCase().includes(companyName.toUpperCase()),
    )
    onFilterUpdate(filteredCompanies)
  }

  return (
    <div>
      <input type="text" onChange={filterByName} value={filterCompanyName} />
    </div>
  )
}

export default MyCompaniesFilters
