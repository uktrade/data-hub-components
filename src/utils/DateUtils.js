import moment from 'moment'

const DATE_FORMAT_LONG = 'YYYY-MM-DD'
const DATE_FORMAT_SHORT = 'YYYY-MM'

function padZero(value) {
  const parsedValue = parseInt(value, 10)
  if (Number.isNaN(parsedValue)) {
    return value
  }
  return parsedValue < 10 ? `0${parsedValue}` : parsedValue.toString()
}

function normaliseAndFormatDate(year, month, day) {
  const y = padZero(year)
  const m = padZero(month)
  const yearAndMonth = `${padZero(y)}-${padZero(m)}`
  return day ? `${yearAndMonth}-${padZero(day)}` : yearAndMonth
}

export default class DateUtils {
  static format(dateStr) {
    return moment(dateStr).format('DD MMM YYYY')
  }

  static formatWithTime(dateTimeStr) {
    return moment(dateTimeStr).format('D MMM YYYY, h:mma')
  }

  static isDateValid = (year, month, day, format = DATE_FORMAT_LONG) => {
    const date = normaliseAndFormatDate(year, month, day)
    return moment(date, format, true).isValid()
  }

  static isShortDateValid(year, month) {
    return DateUtils.isDateValid(year, month, null, DATE_FORMAT_SHORT)
  }

  static transformValueForAPI({ year, month, day = 1 }) {
    if (year && month && day) {
      return normaliseAndFormatDate(year, month, day)
    }

    return null
  }
}
