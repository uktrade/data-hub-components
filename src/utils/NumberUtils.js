export default class NumberUtils {
  static decimal(number) {
    if (!number && number !== 0) {
      return null
    }

    const formatter = new Intl.NumberFormat('en-GB', {
      style: 'decimal',
    })

    return formatter.format(number)
  }

  static currencyGBP(number, options = { maximumSignificantDigits: 21 }) {
    if (!number && number !== 0) {
      return null
    }

    const formatter = new Intl.NumberFormat('en-GB', {
      style: 'currency',
      currency: 'GBP',
      ...options,
    })

    return formatter.format(number)
  }

  static currencyUSD(number, options = { maximumSignificantDigits: 21 }) {
    if (!number && number !== 0) {
      return null
    }

    const formatter = new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      ...options,
    })

    return formatter.format(number)
  }

  static roundToSignificantDigits(number, sigDig) {
    if (!number && number !== 0) {
      return null
    }

    return Number(number.toPrecision(sigDig))
  }
}
