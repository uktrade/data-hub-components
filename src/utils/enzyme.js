export const withTargetValue = value => ({
  target: {
    value,
  },
})

/**
 * @function getOptionValueByText
 * @description Finds an `<option>` element with its inner text matching the
 * {text} parameter and returns the value of its `value` attribute.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {String} text - The option text
 */
export const getOptionValueByText = (wrapper, text) => (
  wrapper.find('option')
    .find({ children: text })
    .props()
    .value
)

/**
 * @function changeAndUpdate
 * @description Simulates the `change` event on the provided {wrapper} with
 * {vaule} as `target.value` and updates (re-renders) the wrapper.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {*} value - The value to be used as `event.target`
 */
export const changeAndUpdate = (wrapper, value) => {
  wrapper.simulate('change', withTargetValue(value))
  wrapper.update()
}

/**
 * @function changeSelectAndUpdate
 * @description Finds a `<select>` element in the wrapper, selects the option
 * matching the {text} and updates (re-renders) the wrapper.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {String} text - The option text
 */
export const changeSelectAndUpdate = (wrapper, text) => {
  changeAndUpdate(wrapper.find('select'), getOptionValueByText(wrapper, text))
  wrapper.update()
}
