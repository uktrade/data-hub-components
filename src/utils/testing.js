// eslint-disable-next-line import/prefer-default-export
export const withTargetValue = value => ({
  target: {
    value,
  },
})
