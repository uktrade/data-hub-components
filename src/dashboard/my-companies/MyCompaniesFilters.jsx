import React from 'react'
import styled from 'styled-components'
import { SelectInput } from '@govuk-react/select'
import LabelText from '@govuk-react/label-text'
import { MEDIA_QUERIES, SPACING } from '@govuk-react/constants'
import { typography } from '@govuk-react/lib'
import useMyCompaniesContext from './useMyCompaniesContext'
import { FILTER_CHANGE, ORDER_CHANGE } from './constants'

const StyledSelectInput = styled(SelectInput)(typography.font({ size: 14 }), {
  [MEDIA_QUERIES.LARGESCREEN]: {
    width: 'auto',
  },
})

const StyledLabelText = styled(LabelText)(typography.font({ size: 14 }))

const StyledContainer = styled.div`
  display: flex;
  justify-content: space-between;
`

const StyledSearch = styled.input`
  margin-right: ${SPACING.SCALE_2};
  padding-right: ${SPACING.SCALE_1};
  padding-left: ${SPACING.SCALE_1};
`

function MyCompaniesFilters() {
  const { dispatch } = useMyCompaniesContext()

  return (
    <StyledContainer>
      <StyledSearch
        type="text"
        placeholder="Search list"
        onChange={e => dispatch({
          type: FILTER_CHANGE,
          filter: e.target.value,
        })}
      />
      <StyledLabelText>
        Sort By:{' '}
        <StyledSelectInput
          onChange={e => dispatch({
            type: ORDER_CHANGE,
            sortBy: e.target.value,
          })}
        >
          <option value="recent">Recent interaction</option>
          <option value="least-recent">Least recent interaction</option>
          <option value="alphabetical">Company name A-Z</option>
        </StyledSelectInput>
      </StyledLabelText>
    </StyledContainer>
  )
}

export default MyCompaniesFilters
