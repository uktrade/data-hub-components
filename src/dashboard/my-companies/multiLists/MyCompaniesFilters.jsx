import React, { useRef } from 'react'
import styled from 'styled-components'
import { SelectInput } from '@govuk-react/select'
import LabelText from '@govuk-react/label-text'
import { MEDIA_QUERIES, SPACING } from '@govuk-react/constants'
import { typography } from '@govuk-react/lib'
import ACTIONS from '../constants'
import useMyCompaniesContext from '../useMycompaniesListsContext'

const StyledSelectInput = styled(SelectInput)(typography.font({ size: 14 }), {
  [MEDIA_QUERIES.LARGESCREEN]: {
    width: 'auto',
  },
})

const StyledLabelText = styled(LabelText)(typography.font({ size: 14 }))

const StyledContainer = styled.div`
  display: flex;
  justify-content: space-between;
`

const StyledSearch = styled.input`
  margin-right: ${SPACING.SCALE_2};
  padding-right: ${SPACING.SCALE_1};
  padding-left: ${SPACING.SCALE_1};
`

function MyCompaniesFilters() {
  const { state, dispatch } = useMyCompaniesContext()
  const inputEl = useRef()

  const handleSelectChange = (e) => {
    dispatch({ type: ACTIONS.SORT_BY, sortType: e.target.value })
  }

  const handleInputChange = (e) => {
    dispatch({ type: ACTIONS.FILTER_BY, filterText: e.target.value })
  }

  return (
    <StyledContainer>
      <StyledSearch
        type="text"
        placeholder="Search list"
        onChange={handleInputChange}
        value={state.filterText}
        ref={inputEl}
      />
      <StyledLabelText>
        Sort By:{' '}
        <StyledSelectInput onChange={handleSelectChange} value={state.sortType}>
          <option value="recent">Recent interaction</option>
          <option value="least-recent">Least recent interaction</option>
          <option value="alphabetical">Company name A-Z</option>
        </StyledSelectInput>
      </StyledLabelText>
    </StyledContainer>
  )
}

export default MyCompaniesFilters
