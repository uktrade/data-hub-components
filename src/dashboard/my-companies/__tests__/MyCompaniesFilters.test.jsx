import React from 'react'
import { mount } from 'enzyme'
import MyCompaniesFilters from '../MyCompaniesFilters'
import useMyCompaniesContext from '../useMyCompaniesContext'
import { withTargetValue } from '../../../utils/enzyme'

describe('MyCompaniesFilters', () => {
  describe('Select', () => {
    test('Should update context state on interaction', () => {
      const stateHistory = []

      const SomeComponent = () => {
        const { state: { filter, sortBy } } = useMyCompaniesContext()
        stateHistory.push({ filter, sortBy })
        return null
      }

      const wrapper = mount(
        <useMyCompaniesContext.Provider>
          <MyCompaniesFilters />
          <SomeComponent />
        </useMyCompaniesContext.Provider>,
      )

      wrapper.find('select')
        .simulate('change', withTargetValue('alphabetical'))
        .simulate('change', withTargetValue('least-recent'))
        .simulate('change', withTargetValue('recent'))

      wrapper.find('input')
        .simulate('change', withTargetValue('foo'))
        .simulate('change', withTargetValue('bar'))
        .simulate('change', withTargetValue('baz'))
        .simulate('change', withTargetValue(''))

      expect(stateHistory).toEqual([
        {
          filter: '',
          sortBy: 'recent',
        },
        {
          filter: '',
          sortBy: 'alphabetical',
        },
        {
          filter: '',
          sortBy: 'least-recent',
        },
        {
          filter: '',
          sortBy: 'recent',
        },
        {
          filter: 'foo',
          sortBy: 'recent',
        },
        {
          filter: 'bar',
          sortBy: 'recent',
        },
        {
          filter: 'baz',
          sortBy: 'recent',
        },
        {
          filter: '',
          sortBy: 'recent',
        },
      ])
    })
  })
})
