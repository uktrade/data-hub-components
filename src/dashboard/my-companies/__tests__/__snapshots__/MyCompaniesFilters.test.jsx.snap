// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyCompaniesFilters renders company filters 1`] = `
.c3 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  color: #0b0c0c;
  display: block;
  clear: none;
  padding-bottom: 2px;
  margin-bottom: 0;
}

.c4 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 16px;
  line-height: 1.25;
  box-sizing: border-box;
  width: 100%;
  height: 33px;
  padding: 5px 4px 4px;
  border: 2px solid #0b0c0c;
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 12px;
  line-height: 1.25;
}

.c4:focus {
  outline: 3px solid #ffbf47;
  outline-offset: 0;
}

.c2 {
  font-family: "nta",Arial,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-weight: 400;
  font-size: 12px;
  line-height: 1.25;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  margin-right: 10px;
  padding-right: 5px;
  padding-left: 5px;
}

@media print {
  .c3 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media print {
  .c3 {
    color: #000;
  }
}

@media only screen and (min-width:641px) {
  .c3 {
    margin-bottom: 0;
  }
}

@media print {
  .c4 {
    font-size: 14px;
    line-height: 1.15;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    font-size: 19px;
    line-height: 1.3157894736842106;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    width: 50%;
    height: 38px;
  }
}

@media print {
  .c4 {
    font-size: 12px;
    line-height: 1.2;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    font-size: 14px;
    line-height: 1.4285714285714286;
  }
}

@media only screen and (min-width:641px) {
  .c4 {
    width: auto;
  }
}

@media print {
  .c2 {
    font-size: 12px;
    line-height: 1.2;
  }
}

@media only screen and (min-width:641px) {
  .c2 {
    font-size: 14px;
    line-height: 1.4285714285714286;
  }
}

<Provider>
  <MyCompaniesFilters>
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-gzVnrw",
              "isStatic": true,
              "lastClassName": "c0",
              "rules": Array [
                "
  display: flex;
  justify-content: space-between;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-gzVnrw",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="c0"
        >
          <styled.input
            onChange={[Function]}
            placeholder="Search list"
            type="text"
            value=""
          >
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-htoDjs",
                    "isStatic": true,
                    "lastClassName": "c1",
                    "rules": Array [
                      "
  margin-right: ",
                      "10px",
                      ";
  padding-right: ",
                      "5px",
                      ";
  padding-left: ",
                      "5px",
                      ";
",
                    ],
                  },
                  "displayName": "styled.input",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-htoDjs",
                  "target": "input",
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={
                Object {
                  "current": .c0 {
  margin-right: 10px;
  padding-right: 5px;
  padding-left: 5px;
}

<input
                    class="c0"
                    placeholder="Search list"
                    type="text"
                    value=""
                  />,
                }
              }
              onChange={[Function]}
              placeholder="Search list"
              type="text"
              value=""
            >
              <input
                className="c1"
                onChange={[Function]}
                placeholder="Search list"
                type="text"
                value=""
              />
            </StyledComponent>
          </styled.input>
          <Styled(LabelText)>
            <StyledComponent
              forwardedComponent={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "attrs": Array [],
                  "componentStyle": ComponentStyle {
                    "componentId": "sc-bZQynM",
                    "isStatic": true,
                    "lastClassName": "c2",
                    "rules": Array [
                      "font-family: \\"nta\\", Arial, sans-serif; -webkit-font-smoothing: antialiased; -moz-osx-font-smoothing: grayscale; @media print {
  font-size: 12px; line-height: 1.2;
} font-weight: 400; font-size: 12px; line-height: 1.25; @media only screen and (min-width: 641px) {
  font-size: 14px; line-height: 1.4285714285714286;
}",
                    ],
                  },
                  "displayName": "Styled(LabelText)",
                  "foldedComponentIds": Array [],
                  "render": [Function],
                  "styledComponentId": "sc-bZQynM",
                  "target": [Function],
                  "toString": [Function],
                  "warnTooManyClasses": [Function],
                  "withComponent": [Function],
                }
              }
              forwardedRef={null}
            >
              <LabelText
                className="c2"
              >
                <styled.span
                  className="c2"
                >
                  <StyledComponent
                    className="c2"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-bwzfXH",
                          "isStatic": false,
                          "lastClassName": "c3",
                          "rules": Array [
                            "font-family: \\"nta\\", Arial, sans-serif; -webkit-font-smoothing: antialiased; -moz-osx-font-smoothing: grayscale; @media print {
  font-size: 14px; line-height: 1.15;
} font-weight: 400; font-size: 16px; line-height: 1.25; @media only screen and (min-width: 641px) {
  font-size: 19px; line-height: 1.3157894736842106;
}",
                            "color: #0b0c0c; @media print {
  color: #000;
}",
                            "display: block; clear: none; padding-bottom: 2px;",
                            [Function],
                            [Function],
                          ],
                        },
                        "displayName": "styled.span",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-bwzfXH",
                        "target": "span",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                  >
                    <span
                      className="c2 c3"
                    >
                      Sort By:
                       
                      <Styled(styled.select)
                        onChange={[Function]}
                        value="recent"
                      >
                        <StyledComponent
                          forwardedComponent={
                            Object {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": Array [],
                              "componentStyle": ComponentStyle {
                                "componentId": "sc-EHOje",
                                "isStatic": false,
                                "lastClassName": "c4",
                                "rules": Array [
                                  "font-family: \\"nta\\", Arial, sans-serif; -webkit-font-smoothing: antialiased; -moz-osx-font-smoothing: grayscale; @media print {
  font-size: 14px; line-height: 1.15;
} font-weight: 400; font-size: 16px; line-height: 1.25; @media only screen and (min-width: 641px) {
  font-size: 19px; line-height: 1.3157894736842106;
}",
                                  "box-sizing: border-box; width: 100%; height: 33px; padding: 5px 4px 4px; border: 2px solid #0b0c0c; @media only screen and (min-width: 641px) {
  width: 50%; height: 38px;
} :focus {
  outline: 3px solid #ffbf47; outline-offset: 0;
}",
                                  [Function],
                                  "font-family: \\"nta\\", Arial, sans-serif; -webkit-font-smoothing: antialiased; -moz-osx-font-smoothing: grayscale; @media print {
  font-size: 12px; line-height: 1.2;
} font-weight: 400; font-size: 12px; line-height: 1.25; @media only screen and (min-width: 641px) {
  font-size: 14px; line-height: 1.4285714285714286;
}",
                                  "@media only screen and (min-width: 641px) {
  width: auto;
}",
                                ],
                              },
                              "displayName": "Styled(styled.select)",
                              "foldedComponentIds": Array [
                                "sc-ifAKCX",
                              ],
                              "render": [Function],
                              "styledComponentId": "sc-EHOje",
                              "target": "select",
                              "toString": [Function],
                              "warnTooManyClasses": [Function],
                              "withComponent": [Function],
                            }
                          }
                          forwardedRef={null}
                          onChange={[Function]}
                          value="recent"
                        >
                          <select
                            className="c4"
                            onChange={[Function]}
                            value="recent"
                          >
                            <option
                              value="recent"
                            >
                              Recent interaction
                            </option>
                            <option
                              value="least-recent"
                            >
                              Least recent interaction
                            </option>
                            <option
                              value="alphabetical"
                            >
                              Company name A-Z
                            </option>
                          </select>
                        </StyledComponent>
                      </Styled(styled.select)>
                    </span>
                  </StyledComponent>
                </styled.span>
              </LabelText>
            </StyledComponent>
          </Styled(LabelText)>
        </div>
      </StyledComponent>
    </styled.div>
  </MyCompaniesFilters>
</Provider>
`;
